server/src/main/java/org/opensearch/common/Rounding.java

```diff
@@ -1365,4 +1387,64 @@ public double roundingSize(long utcMillis, DateTimeUnit timeUnit) {
             return delegate.roundingSize(utcMillis, timeUnit);
         }
     }
+
+    /**
+     * Implementation of {@link Prepared} using pre-calculated "round down" points.
+     *
+     * <p>
+     * It uses linear search to find the greatest round-down point less than or equal to the given timestamp.
+     * For small inputs (&le; 64 elements), this can be much faster than binary search as it avoids the penalty of
+     * branch mispredictions and pipeline stalls, and accesses memory sequentially.
+     *
+     * <p>
+     * It uses "meet in the middle" linear search to avoid the worst case scenario when the desired element is present
+     * at either side of the array. This is helpful for time-series data where velocity increases over time, so more
+     * documents are likely to find a greater timestamp which is likely to be present on the right end of the array.
+     *
+     * @opensearch.internal
+     */
+    @InternalApi
+    static class BidirectionalLinearSearchArrayRounding implements Prepared {
+        private final long[] ascending;
+        private final long[] descending;
+        private final Prepared delegate;
+
+        BidirectionalLinearSearchArrayRounding(long[] values, int max, Prepared delegate) {
+            assert max > 0 : "at least one round-down point must be present";
+            this.delegate = delegate;
+            int len = (max + 1) >>> 1; // rounded-up to handle odd number of values
+            ascending = new long[len];
+            descending = new long[len];
+
+            for (int i = 0; i < len; i++) {
+                ascending[i] = values[i];
+                descending[i] = values[max - i - 1];
+            }
+        }
+
+        @Override
+        public long round(long utcMillis) {
+            int i = 0;
+            for (; i < ascending.length; i++) {
+                if (descending[i] <= utcMillis) {
```

Comment:
@ketanv3 just a question please

> This is ideal for sufficiently large arrays. But a simple linear search is far superior for small arrays as it avoids the penalty of branch misprediction and pipeline stalls, and accesses memory sequentially.

Wouldn't it be more efficient to use a single array instead and not jump between two arrays (essentially, two memory locations) instead? It seems like we could do it easiy:
 - allocate one array of 2 * len
 - user 0..len for asscessing, use len + 1 .. 2 * len for descessing

Wdyt?

Reply:
@reta Glad you asked. I tried three variants of this:
1. Keep a single sorted array and access it with `i` and `length - i - i` for ascending and descending values.
2. Keep a single sorted array and access it with two separate pointers.
3. Keep a single array with ascending and descending halves as you suggested, and access `i` and `mid + i` for ascending and descending values.

They performed poorly compared to the current implementation. Reasons being, they involve more operations in the hot path (either assigning more variables or involving an extra addition on subtraction operation).

Reply:
Despite, (3) wouldn't be much different from the current implementation from a memory jump point of view. `i` and `mid + i` would be similar distance apart compared to the current implementation (in most allocations).